_posts/2019-04-24-accessing-kafka-part-2.md

```diff
@@ -0,0 +1,317 @@
+```
+
+The node address and node port number give you all what is needed to connect to your cluster:
+
+```
+bin/kafka-console-producer.sh --broker-list <node-address>:<node-port> --topic <your-topic>
+```
+
+Fore more details, see the [Strimzi documentation](https://strimzi.io/docs/latest/full.html#proc-accessing-kafka-using-loadbalancers-deployment-configuration-kafka).
+
+# Troubleshooting node ports
+
+Node ports are fairly complex to configure.
+There are many things which can go wrong.
+
+One of the common problems is that the address which is presented to Strimzi by the Kubernetes API in the node resource is not accessible from outside.
+This can happen for example because the DNS name or IP address used there is only internal and cannot be reached by the client.
+This might happen with production grade cluster but also with local development tools such as Minikube or Minishift.
+In such case, you might get following errors from your client:
+
+```
+[2019-04-22 21:04:11,976] WARN [Consumer clientId=consumer-1, groupId=console-consumer-42133] Connection to node 1 (/10.0.2.15:31301) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
+```
+
+or
+
+```
+[2019-04-22 21:11:37,295] WARN [Producer clientId=console-producer] Connection to node -1 (/10.0.2.15:31488) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
+```
+
+When you see one of the errors, you can compare the following addresses:
+* The address under which you expect your nodes to be reachable by your Kafka clients (with Minikube, that should be the IP address returned by the `minikube ip` command).
+* The address of the nodes where the Kafka pods are running (`kubectl get node <node-name> -o=jsonpath='{range .status.addresses[*]}{.type}{"\t"}{.address}{"\n"}'`
+* The address advertised by the Kafka broker (`kubectl exec my-cluster-kafka-0 -c kafka -it -- cat /tmp/strimzi.properties | grep advertised`)
+
+If these addresses are different, they are most probably causing your problem.
+Strimzi can help you with this problem.
+You can use the override options to change the addresses advertised by the Kafka pods.
+Instead of reading the address from the Kubernetes API from the node resources, you can configure them in the `Kafka` custom resource.
+For example:
+
+```yaml
+# ...
+listeners:
+  external:
+    type: nodeport
+    tls: false
+    overrides:
+      brokers:
+      - broker: 0
+        advertisedHost: XXX.XXX.XXX.XXX
+      - broker: 1
+        advertisedHost: XXX.XXX.XXX.XXX
+      - broker: 2
+        advertisedHost: XXX.XXX.XXX.XXX
+# ...
+```
+
+The overrides can specify either a DNS names or IP addresses.
+This solution is not optimal, since you would need to maintain the addresses in the custom resource and remember to update them every time you are for example scaling up your Kafka cluster.
```

Comment:
```suggestion
This solution is not ideal, since you need to maintain the addresses in the custom resource and remember to update them every time you are, for example, scaling up your Kafka cluster.
```